File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
void generateParenthesis(int open, int close, string op, vector<string>& ans) {
        if(open == 0 && close == 0) {
            ans.push_back(op);
            return;
        }
        
        if(open > 0) generateParenthesis(open-1, close, op+"(", ans);
        
        if(close > open) generateParenthesis(open, close-1, op+")", ans);
        
    }
    
    vector<string> generateParenthesis(int n) {
         vector<string> ans;
        int open = n;
        int close = n;
        string op = "";
        generateParenthesis(open, close, op, ans);
        
        return ans;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        ans = []
        
        def dfsol(open,close,st,ans):
            if (open==0 and close ==0):
                ans.append(st)
                return
            
            if open>0:
                dfsol(open-1,close,st+"(",ans)
            
            if open<close:
                dfsol(open,close-1,st+")",ans)
        
        
        dfsol(n,n,"",ans)
        
        return ans
        